pkg/steps/run_test.go

```diff
@@ -40,135 +44,282 @@ func (*fakeStep) Objects() []ctrlruntimeclient.Object { return nil }
-		if step.shouldRun && step.numRuns != 1 {
-			t.Errorf("step %s did not run just once, but %d times", step.name, step.numRuns)
-		}
-		if !step.shouldRun && step.numRuns != 0 {
-			t.Errorf("step %s expected to never run, but ran %d times", step.name, step.numRuns)
-		}
-	}
-}
+			var steps []api.Step
+			for _, step := range tc.steps {
+				if step.runErr != nil {
+					shouldFail++
+				}
+				if step.shouldRun {
+					shouldRun++
+				}
+				steps = append(steps, step)
+			}
 
-func TestRunFailureCase(t *testing.T) {
-	root := &fakeStep{
-		name:      "root",
-		shouldRun: true,
-		requires:  []api.StepLink{api.ExternalImageLink(api.ImageStreamTagReference{Namespace: "ns", Name: "base", Tag: "latest"})},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceRoot)},
-	}
-	other := &fakeStep{
-		name:      "other",
-		shouldRun: true,
-		requires:  []api.StepLink{api.ExternalImageLink(api.ImageStreamTagReference{Namespace: "ns", Name: "base", Tag: "other"})},
-		creates:   []api.StepLink{api.InternalImageLink("other")},
-	}
-	src := &fakeStep{
-		name:      "src",
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceRoot)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceSource)},
-	}
-	bin := &fakeStep{
-		name:      "bin",
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceSource)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceBinaries)},
-	}
-	testBin := &fakeStep{
-		name:      "testBin",
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceSource)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceTestBinaries)},
-	}
-	rpm := &fakeStep{
-		name:      "rpm",
-		runErr:    errors.New("oopsie"),
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceBinaries)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceRPMs)},
-	}
-	unrelated := &fakeStep{
-		name:      "unrelated",
-		shouldRun: false,
-		requires:  []api.StepLink{api.InternalImageLink("other"), api.InternalImageLink(api.PipelineImageStreamTagReferenceRPMs)},
-		creates:   []api.StepLink{api.InternalImageLink("unrelated")},
-	}
-	final := &fakeStep{
-		name:      "final",
-		shouldRun: false,
-		requires:  []api.StepLink{api.InternalImageLink("unrelated")},
-		creates:   []api.StepLink{api.InternalImageLink("final")},
-	}
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			if tc.cancelled {
+				cancel()
+			}
+			suites, _, errs := Run(ctx, api.BuildGraph(steps))
+			if errs == nil && len(tc.errExpected) > 0 {
+				t.Error("got no error but expected one")
+			}
+			actualErrors := results.MapMessagesByReason(results.ConvertToErrorList(errs))
```

Comment:
Can't we simply do this via a string instead if two funcs that need to be used in conjunction and require some weird data structure as input? Something like `func DeterministicErrorMsg(errs []error) (string, error)`

Reply:
I don't want to compare only the error messages, but the reason too.

Reply:
And the reason is not part of the error string or what?

Reply:
We can have different messages for the same reason. That's why I am using a `map[string][]string`

Reply:
I still don't follow, sorry, So the error message only includes one of `Reason` and `Other error details`? Otherwise you could simply string sort

Reply:
the error message doesn't include any `Reason`. This is another field of the struct.

Reply:
Ah, can you then just iterate through all untyped errors, type assert, if match store a a string of `reason: message` in a string slice, if not just the message, sort the slice and join with `\n`? That would make using this significantly easier

Reply:
This is exactly what I did in the first version of this PR :P

Reply:
nevermind. I reverted it back. Can you check?